leaderelection/src/main/java/io/scalecube/leaderelection/RaftLeaderElection.java

```diff
@@ -0,0 +1,423 @@
+import io.scalecube.transport.Address;
+import io.scalecube.transport.Message;
+
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import rx.Observable;
+import rx.schedulers.Schedulers;
+import rx.subjects.PublishSubject;
+import rx.subjects.Subject;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.ThreadLocalRandom;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+
+/**
+ * Raft nodes are always in one of three states: follower, candidate, or leader. All nodes initially start out as a
+ * follower. In this state, nodes can cast votes. If no entries are received for some time, nodes self-promote to the
+ * candidate state. In the candidate state, nodes request votes from their peers. If a candidate receives a quorum of
+ * votes, then it is promoted to a leader. Consensus is fault-tolerant up to the point where quorum is available. If a
+ * quorum of nodes is unavailable, it is impossible to reason about peer membership. For example, suppose there are only
+ * 2 peers: A and B. The quorum size is also 2. If either A or B fails, it is now impossible to reach quorum. At this
+ * point, the algorithm uses random timer timeout and will resolve the situation by one of the nodes taking and
+ * maintaining leadership using heartbeats so in this specific case leaders will cancel one another until one of the
+ * nodes sends the hearbeats first raft leader election algorithm: when a node starts it becomes a follower and set a
+ * timer that will trigger after x amount of time and waiting during this time for a leader heartbeats if no leader send
+ * heartbeats then the timer is triggered and node transition to a candidate state - once in a candidate state the node
+ * gossip vote request to all member nodes nodes that receive a vote request will answer directly to the requesting node
+ * with their vote. the candidate starts to collect the votes and check if there is consensus among the cluster members
+ * and that means (N/2 +1) votes once the candidate gained consensus it immediately take leadership and start to gossip
+ * heartbeats to all cluster nodes. a leader node that receives an heartbeat from any other member transition to
+ * follower state in the case where several candidates are running for leadership they both cancel the candidate state
+ * and next timeout node will become a leader and election process is restarted until there is clear consensus among the
+ * cluster https://raft.github.io/
+ * 
+ * @author Ronen Nachmias on 9/11/2016.
+ */
+public class RaftLeaderElection implements LeaderElection {
```

Comment:
There is a 'term' property in raft leader election algorithm which is used to resolve leader in case of network partition, I don't see something which would look like this here?